source/common/http/conn_manager_impl.cc

```diff
@@ -688,6 +688,14 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
 
   decodeHeaders(nullptr, *request_headers_, end_stream);
 
+  if (Grpc::Common::hasGrpcContentType(*request_headers_) && stream_idle_timer_ != nullptr) {
```

Comment:
grpc-timeout control is a router thing, not a base level stream thing. I don't think this change as written is acceptable from a security perspective as we already discussed in the related issue. I think at minimum it needs to be opt-in via config. Additionally, I wonder if the base stream level idle timeout should be optionally populated from grpc-timeout much like we do in the router?

Reply:
Also I am curious how does it work for long running stream? For example xDS connections, we have disabled stream timeouts by setting it to zero - with this feature, do we have to set the max-grpc-timeout and not set stream-idle timeout? If yes, would it not be a breaking change?

Reply:
@mattklein123 Could you elaborate what is your security perspective concern here? AFAIK each gRPC request is bounded by a request timeout (although the bound can be set to unlimited via an explicit opt-in via max-grpc-timeout set to 0). Also what would be the value of setting the idle timeout to grpc-timeout if the request timeout is going to be set to the same value anyway?

Even if the request takes a route without explicitly configured max-grpc-timeout, the request timeout of that route is in effect.

@ramaraochavali I take your question was to Matt, correct me if I was mistaken!

Reply:
> Could you elaborate what is your security perspective concern here?

It's the same concern as every other time we talk about this topic. :) Basically, this change allows a potentially untrusted request to override a timeout, which at the edge is not safe. This timeout is a base request timeout that is important to guard against trickle requests and requests that may not reach the router for various reasons (filters, etc.).

I think there are two issues here:
1) This behavior should be opt-in based on config IMO (maybe we can discuss whether we do this automatically based on internal/external status, but I don't feel great about that either)
2) Whether we should do any automatic timer setup in the base HCM vs. the router, based on gRPC headers.

Reply:
OK, I did not consider the case where the request may not reach the router, and from that viewpoint it makes sense to move this logic to the router. However, we do stuff like override HCM level timeouts with router specified timeouts already now, I wonder why that is not a concern.

One the overall safety topic I feel that the max-grpc-timeout being an opt-in is enough. Or are you concerned when a max-grpc-timeout is not configured, the HCM or route level request timeout could be unsafe, if it overrides the idle-timeout for gRPC requests?

Reply:
The timeout we are looking at here is the base request timeout/idle timeout. This is different from the router timeouts which are upstream timeouts. This timeout is essentially protecting against downtream while the router timeouts are protecting against upstream (with the exception of the idle timeout which is for both downstream and upstream purposes).

So in that sense the timeouts are related but different. I think it's reasonable to allow the base stream idle timeout to be set by headers, but IMO it must be an opt-in config option which defaults to off.

Reply:
@mattklein123 How about re-using the route-level max-grpc-timeout for this at the HCM level? This would obviously be limited to cases where the (cached) route can be found, but essentially we would reset the idle timeout (if not disabled) to the value from the `grpc-timeout` header, but only if a route-level `max-grpc-timeout` is specified (thus the "opt-in") for the cached route? I.e., no change if a cached route is not found with a 'max-grpc-timeout` value explicitly configured. This idle timeout would be in effect even if the request never makes it to the router filter.

Reply:
This seems like a reasonable proposal to me. @alyssawilk any concerns? My only concern is that someone may want a different timeout to be configurable at the base stream layer.

Reply:
Sanity check to make sure I have the background correct.
Right now we have an overall response timeout which is generally per-route, but one can allow a (bounded) client-driven grpc header to override the overall timeout on a per route basis.

The suggestion here is where Envoy users have opted into allowing the client to set a grpc-specific per-stream timeout, that we automatically remove any keepalive timeout to ensure the stream gets the full grpc-timeout duration to make forward progress?

I don't think that is going to make sense as a solution for all users.  In my experience overall timeouts are high bounds for "did we actually complete this request/rpc in a sane amount of time" and idle timeouts (like the stream idle timer) are much shorter and more of a "hey, did this user just suddenly switch networks or close their laptop because if so I want to reclaim my resources"

Using the same value for both may make sense for some use cases where there's known to be a long hang (gRPC equivalent of browser-channel) but not for others, so I'm concerned about assuming it's the right thing for all users of max-grpc-timeout.

That said, adjusting the request_timer_ up to the per-route capped grpc-timeout value does totally make sense if you want to start with that.

Reply:
@alyssawilk We already adjust the request timeout to the value of the `grpc-timeout` for gRPC requests if route has `max-grpc-timeout` explicitly configured. For this to work as expected for gRPC streaming (where longish idle periods may be expected), the route must have idle timeout disabled. As you generally want to use idle timeouts for non-gRPC requests this currently requires use of different routes for gRPC and non-gRPC traffic.

When deploying Envoy as a "transparent" proxy all such app-specific routing configuration is highly undesirable. It seems to me that deferring to the `grpc-timeout` header would allow a transparent proxy to not break the expected gRPC semantics while at the same time operating under reasonable idle timeout for traffic that has no explicit timeout semantic (currently all non-gRPC traffic).

Reply:
Yep, I have no problem with having a configuration option where one can disable the idle timeout for gRPC, I'm just indicating that just because this is your desired behavior does not mean it's everyone else's desired behavior.  Overall timeouts and idle timeouts are substantially different and while some users may want to defer to gRPC for both, I can think of at least one user who does not (but we can work around it on our own, so I don't insist on 2 knobs for Google's sake)

Happy to defer to @mattklein123 and or @lizan if we want to switch over the default behavior or add a second knob.

Reply:
@alyssawilk I'm also fine adding a second knob, but I did not see the need for it yet, do you care to elaborate the use case you referred to?

Reply:
I guess the problem in not configuring an idle timeout is - @jrajahalme wants to configure idle timeout for h1 traffic and not configure for gRPC and rely on max-grpc-timeout that allows him to reuse the route. @jrajahalme is that correct?

Reply:
I agree though that `idle timeout` and `max-grpc-timeout ` are fundamentally different. gRPC connectivity semantics seems to talk about idleness though https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md?

Reply:
> I don't think this is true. In fact, a stream idle timeout is required for safely offering edge streaming APIs to untrusted clients. This is really the crux of my concern; we need to handle the untrusted case in addition to the trusted case.

I think the argument for gRPC here is that gRPC tends to set idle timeout at channel level rather than stream level, so as long as the connection (channel) is not idle timed out, the streams shouldn't be affected. The same semantics does apply to H2 that provides connection level keep-alive mechanism but not at stream level. So I agree this need to be opt-in if you want to change the behavior of gRPC.

> Thinking about this more, if you don't want an idle timeout can't you just not configure one? In which case is this change needed?

Yeah I'd like to know more about the use case that the route level configuration doesn't work.